### PR TITLE
Search bar

### PR DESCRIPTION
**Pull request by fcescob**
This PR adds a SearchBar component and a Navbar to the App, it also implements a "search" screen with some random results. 

## Current Issues

There is an issue with mocking, so while search results should be relevant to the string searched, only one video repeated multiple times appears.

This PR builds off #1248 so any comments concerning Icons should be directed there.

### CODE REVIEW

File: atlas/packages/app/src/components/LayoutWithRouting.tsx [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
-import { HomeView, VideoView } from '@/views'
+import { Navbar } from '@/components'
+import { HomeView, VideoView, SearchView } from '@/views'
 import routes from '@/config/routes'
 import { Router } from '@reach/router'
 
 const LayoutWithRouting: React.FC = () => (
   <main>
     <GlobalStyle />
+    <Router primary>
```
Comment (kdembler) [inline atlas/packages/app/src/components/LayoutWithRouting.tsx:11]: Why do we need this `<Router>`?

File: atlas/packages/app/src/components/Navbar.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import React, { useState } from 'react'
+import styled from '@emotion/styled'
+import { animated, useSpring } from 'react-spring'
+import { ReactComponent as UnstyledLogo } from '../assets/logo.svg'
+import { Searchbar, Icon } from '@/shared/components'
+import { colors } from '@/shared/theme'
+import { navigate, Link, RouteComponentProps } from '@reach/router'
+
+const Header = styled.header<{ isSearching: boolean }>`
```
Comment (kdembler) [inline atlas/packages/app/src/components/Navbar.tsx]: There's a lot of `styled` components here, let's move them to a separate file to keep the code readable

File: atlas/packages/app/src/components/Navbar.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  align-items: center;
+  justify-self: end;
+  :hover {
+    cursor: pointer;
+  }
+`
+
+const Navbar: React.FC<RouteComponentProps> = ({ location }) => {
+  const [search, setSearch] = useState('')
+  const [isSearching, setIsSearching] = useState(Boolean(search.trim()))
```
Comment (kdembler) [inline atlas/packages/app/src/components/Navbar.tsx]: Why `useState(Boolean(search.trim()))`? This will always be `false` - this is only the default value

File: atlas/packages/app/src/components/Navbar.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  const [props, set] = useSpring(() => ({
```
Comment (kdembler) [inline atlas/packages/app/src/components/Navbar.tsx]: Please use more descriptive variable names

Comment (kdembler) [reply atlas/packages/app/src/components/Navbar.tsx]: Also, there is a warning that `props` is never used, what's the purpose of this then?

File: atlas/packages/app/src/components/Navbar.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+          </Link>
+          <Link to="/discover">
+            <StyledIcon icon="binocular" />
+          </Link>
+        </NavigationContainer>
+      )}
+
+      <StyledSearchbar
+        placeholder="Search..."
+        onChange={(e) => {
```
Comment (kdembler) [inline atlas/packages/app/src/components/Navbar.tsx]: It'd be good to move all of this handlers to functions outside of `return` to make the JSX more readable

File: atlas/packages/app/src/components/VideoBestMatch.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+type BestVideoMatchProps = {
+  video: VideoFields
+  onClick: (e: React.MouseEvent<HTMLImageElement>) => void
+}
+const Container = styled.div`
+  color: ${colors.gray[300]};
+  padding-right: 2rem;
+`
+const Header = styled.h3``
```
Comment (kdembler) [inline atlas/packages/app/src/components/VideoBestMatch.tsx]: We don't need that

File: atlas/packages/app/src/components/VideoBestMatch.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import { Button } from '@/shared/components'
+import { formatNumber } from '@/utils/number'
+import { formatDate } from '@/utils/time'
+import { colors } from '@/shared/theme'
+
+type BestVideoMatchProps = {
+  video: VideoFields
+  onClick: (e: React.MouseEvent<HTMLImageElement>) => void
+}
+const Container = styled.div`
```
Comment (kdembler) [inline atlas/packages/app/src/components/VideoBestMatch.tsx]: Like in the `Navbar`, let's move those to a separate file

File: atlas/packages/app/src/components/VideoBestMatch.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      <InnerContainer>
+        <TitleContainer>
+          <Title>{title}</Title>
+          <span>
+            {formatNumber(views)} views • {formatDate(publishedOnJoystreamAt)}
+          </span>
+        </TitleContainer>
+        <ButtonsGroup>
+          <Button variant="secondary">Share</Button>
+          <Button icon="block" />
```
Comment (kdembler) [inline atlas/packages/app/src/components/VideoBestMatch.tsx]: Do we need those? I know they're part of the design but I don't think we'll be using them for now? Same for the share button

File: atlas/packages/app/src/components/VideoGallery.tsx [hunk 9]
```diff
@@ -41,7 +41,9 @@ const VideoGallery: React.FC<VideoGalleryProps> = ({ title, action, videos, load
   }, [])
 
   const handleVideoClick = (id: string) => {
-    navigate(routes.video(id))
+    navigate(routes.video(id), {
+      replace: true,
```
Comment (kdembler) [inline atlas/packages/app/src/components/VideoGallery.tsx]: Why this change? If a user goes to video view from the landing page, we want them to be able to go back, right?

File: atlas/packages/app/src/mocking/server.ts [hunk 10]
```diff
@@ -33,6 +34,38 @@ createServer({
...
             return shuffle(channels)
           },
+          // FIXME: This resolver is currently broken and returns the same result n times instead of the correct result.
+          search: (obj: unknown, { query_string }: SearchVariables, context: unknown, info: unknown) => {
+            const items = [...mockVideos, ...mockChannels]
+
+            let rankCount = 0
+            const matchQueryStr = (str: string) => str.includes(query_string) || query_string.includes(str)
+
+            const relevantItems = items.flatMap((item: any) => {
```
Comment (kdembler) [inline atlas/packages/app/src/mocking/server.ts]: This is overly complicated and hard to read with all the nesting. Also, when you need to build upon an array on some conditions it may make more sense to use `reduce` instead of `flatMap` and returning `[]`.

Consider the following example:
```tsx
const relevantItems = items.reduce((acc, item) => {
  const matched = item.__typename === 'Channel' ? matchQueryStr(item.handle) : matchQueryStr(item.description) || matchQueryStr(item.title)
  if (!matched) return
  return [
    ...acc,
    {
      __typename: 'FreeTextSearchResult',
      item,
      rank: rankCount++,
    }
  ]
}, [])
```

File: atlas/packages/app/src/views/SearchView.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  const { data, loading } = useQuery<Search, SearchVariables>(SEARCH, { variables: { query_string: search } })
+
+  if (loading || !data) {
+    return <p>Loading...</p>
+  }
+  if (!data.search) {
+    return <p>Something went wrong...</p>
+  }
+
+  const results = data.search.slice().sort((a, b) => a.rank - b.rank)
```
Comment (kdembler) [inline atlas/packages/app/src/views/SearchView.tsx]: This and the following 2 lines are potentially expensive and we shouldn't do them on every render. Consider using memoization here

File: atlas/packages/app/src/components/LayoutWithRouting.tsx (hunk 1 above)
Comment (kdembler) [reply atlas/packages/app/src/components/LayoutWithRouting.tsx:11]: Is this because of navigation being done inside the navbar?

**Review by kdembler — CHANGES_REQUESTED** [review]
Apart from the comments, some things that will need improvement:

### UI
- We should indicate no search results, currently if there is nothing found, we're left with an empty screen with just the tabs
- The navbar changes its height when focusing/closing the searchbar, I don't think that's desired
- Should we present videos search result in a carousel? We have a lot of screen space, that seems like a weird choice. Especially in the videos tab when there is nothing else to browse. @bedeho 
- Navigation icons could use some more contrast - it's hard to spot them when they're this gray

### Code
- Please mind the warnings outputted by ESLint, there's a number of unused variables/imports
- Sometimes it's fine to have styled components alongside the main component but if there's a lot of them, we should keep them separate 
- You're repeatedly using raw pixel values. Shouldn't we use spacing values from the theme instead?

File: atlas/packages/app/src/mocking/server.ts (hunk 10 above)
Comment (fcescob) [reply atlas/packages/app/src/mocking/server.ts]: This looks cleaner indeed, changed it.

File: atlas/packages/app/src/components/VideoGallery.tsx (hunk 9 above)
Comment (fcescob) [reply atlas/packages/app/src/components/VideoGallery.tsx]: I forgot to remove that, removed it now.

File: atlas/packages/app/src/components/VideoBestMatch.tsx (hunk 8 above)
Comment (fcescob) [reply atlas/packages/app/src/components/VideoBestMatch.tsx]: Removed them

File: atlas/packages/app/src/components/VideoBestMatch.tsx (hunk 7 above)
Comment (fcescob) [reply atlas/packages/app/src/components/VideoBestMatch.tsx]: Done!

File: atlas/packages/app/src/components/Navbar.tsx (hunk 5 above)
Comment (fcescob) [reply atlas/packages/app/src/components/Navbar.tsx]: Done!

File: atlas/packages/app/src/components/VideoBestMatch.tsx (hunk 6 above)
Comment (fcescob) [reply atlas/packages/app/src/components/VideoBestMatch.tsx]: Removed

File: atlas/packages/app/src/components/Navbar.tsx (hunk 4 above)
Comment (fcescob) [reply atlas/packages/app/src/components/Navbar.tsx]: It was used in an old animated version, removed it for now.

File: atlas/packages/app/src/components/Navbar.tsx (hunk 3 above)
Comment (fcescob) [reply atlas/packages/app/src/components/Navbar.tsx]: Thanks for catching that, replaced that with `false`

File: atlas/packages/app/src/components/Navbar.tsx (hunk 2 above)
Comment (fcescob) [reply atlas/packages/app/src/components/Navbar.tsx]: Done!

**Comment by fcescob** [comment]
> Apart from the comments, some things that will need improvement:
> 
> ### UI
> * We should indicate no search results, currently if there is nothing found, we're left with an empty screen with just the tabs
> * The navbar changes its height when focusing/closing the searchbar, I don't think that's desired
> * Should we present videos search result in a carousel? We have a lot of screen space, that seems like a weird choice. Especially in the videos tab when there is nothing else to browse. @bedeho
> * Navigation icons could use some more contrast - it's hard to spot them when they're this gray


Regarding this:

- Let's handle this in #1289 
- This is part of the design, the height of the navbar goes from `72px` to `64px` when focused
- This is a good point, I will change that to be like in the `VideoView` , I've extracted a `VideoGrid` component and used that. 
- I used `gray[600]` for the icons as that is what the arrow icons where using in the design, so I think this is the intended contrast for icons on the navbar as specified by the design, but we can open an issue about it and discuss if we feel like it makes them hard to see.

File: atlas/packages/app/src/components/Navbar/Navbar.tsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  const [search, setSearch] = useState('')
+  const [isSearching, setIsSearching] = useState(false)
+
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setSearch(e.currentTarget.value)
+  }
+
+  const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter' || (e.key === 'NumpadEnter' && search.trim() !== '')) {
+      navigate(`/search/${search}`)
```
Comment (kdembler) [inline atlas/packages/app/src/components/Navbar/Navbar.tsx]: Let's use the value from `routes` here

File: atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  onBlur,
+  onSubmit,
+  ...hmtlProps
+}) => {
+  return (
+    <Input
+      value={value}
+      placeholder={placeholder}
+      type="search"
+      autoSave="some_unique_value"
```
Comment (kdembler) [inline atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx]: What is that value?

File: atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      name="s"
```
Comment (kdembler) [inline atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx]: Why `s`?

**Review by kdembler — COMMENTED** [review]
Also, there's still this: `You're repeatedly using raw pixel values. Shouldn't we use spacing values from the theme instead?`. What's your take on that?

**Comment by kdembler** [comment]
One more comment to what you wrote:
> I used gray[600] for the icons as that is what the arrow icons where using in the design

That's actually not true. The design uses this color for a disabled arrow, when you cannot navigate forward. The go back arrow has white color, which we should use for the navigation icons

File: atlas/packages/app/src/components/Navbar/Navbar.tsx (hunk 12 above)
Comment (fcescob) [reply atlas/packages/app/src/components/Navbar/Navbar.tsx]: Good catch thanks!

File: atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx (hunk 13 above)
Comment (fcescob) [reply atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx]: `autoSave` is a non standard attribute that's supposed to persist the input string between page refreshes, I had seen it and it looked useful, so I've added it, thinking about it, this is useless with React. So I'm removing it now.

https://www.wufoo.com/html5/autosave-attribute/

File: atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx (hunk 14 above)
Comment (fcescob) [reply atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx]: Removed thanks.

**Comment by fcescob** [comment]
> One more comment to what you wrote:
> 
> > I used gray[600] for the icons as that is what the arrow icons where using in the design
> 
> That's actually not true. The design uses this color for a disabled arrow, when you cannot navigate forward. The go back arrow has white color, which we should use for the navigation icons

I thought this was the case since the other clickable elements like the searchbar  had a similar contrast, and I assumed the navigation arrows being white as being focused or hovered. Anyway more contrast never hurts, so I'm moving them to be `gray[300]` like the non selected text in the tabs.

File: atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx (hunk 13 above)
Comment (kdembler) [reply atlas/packages/app/src/shared/components/Searchbar/Searchbar.tsx]: @Gamaranto I think this isn't removed

**Review by kdembler — APPROVED** [review]
LGTM

**Review by bedeho — APPROVED** [review]
